### PR TITLE
Firebase authentication provider is implemented

### PR DESCRIPTION
**Pull request by mockitor**

### CODE REVIEW

File: src/authDataManager/firebase.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Helper functions for accessing the qq Graph API.
```
Comment (flovilmart) [inline src/authDataManager/firebase.js]: bad comment.

File: src/authDataManager/firebase.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+var Parse = require('parse/node').Parse;
+var admin = require("firebase-admin");
+
+
```
Comment (flovilmart) [inline src/authDataManager/firebase.js]: remove empty line

File: src/authDataManager/firebase.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+var firebaseServiceAccount = process.env.FIREBASE_SERVICE_ACCOUNT;
+if (firebaseServiceAccount != null) {
+  admin.initializeApp({
+    credential: admin.credential.cert(require(firebaseServiceAccount))
+  });
+}
+
+
```
Comment (flovilmart) [inline src/authDataManager/firebase.js]: remove empty line

File: src/authDataManager/firebase.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  return admin.auth().verifyIdToken(authData.access_token)
+      .then(function (decodedToken) {
+        if (decodedToken == null || decodedToken.uid == null || decodedToken.uid == "") {
+          throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, "Firebase user not found");
+        }
+        return;
+      }).catch(function (error) {
+        throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, error.message);
+      });
+
```
Comment (flovilmart) [inline src/authDataManager/firebase.js]: remove empty line

File: src/authDataManager/firebase.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// Returns a promise that fulfills if this app id is valid.
+function validateAppId() {
+  return Promise.resolve();
+}
+
+
```
Comment (flovilmart) [inline src/authDataManager/firebase.js]: remove empty line

File: src/authDataManager/index.js [hunk 6]
```diff
@@ -13,6 +13,8 @@ let vkontakte = require("./vkontakte");
 let qq = require("./qq");
 let wechat = require("./wechat");
 let weibo = require("./weibo");
+let firebase = require("./firebase");
+
```
Comment (flovilmart) [inline src/authDataManager/index.js]: remove empty line

File: package.json [hunk 7]
```diff
@@ -23,6 +23,7 @@
     "commander": "2.9.0",
     "deepcopy": "0.6.3",
     "express": "4.14.0",
+    "firebase-admin": "^4.0.3",
```
Comment (flovilmart) [inline package.json]: not sure we want to depend on that module for every single user.

You can probably provide that module externally.

**Comment by flovilmart** [comment]
In general, while this is very interesting, this has a strong dependency on firebase-admin. Which is not required for all users.

We try to minimize the amount of direct dependencies on that repository.

So, probably you would want to make that dependency an optional one or even better, provide this authentication through a 3rd party package that a user can configure:

```
new ParseServer({
  oauth: {
    firebase: {
       module: "parse-firebase-auth"
    }
  }
});
```

See:
https://github.com/ParsePlatform/parse-server/blob/master/src/authDataManager/index.js#L79

I'm planning to refactor the auth layer through proper adapters to match the overall architecture.

File: package.json (hunk 7 above)
Comment (mockitor) [inline package.json]: ok, i'll try to provide externally. thanks for your feedbacks.

**Comment by acinader** [comment]
would be good if i could find an eslint rule for the blank line formatting stuff....

http://eslint.org/docs/rules/no-multiple-empty-lines

that seems too lienent, but having max at 1 would likely be too strict.

**Comment by flovilmart** [comment]
@mockitor We've merged the refactoring for the adapter loading, providing your adapter externally should be as easy as
```
npm install --save firebase-auth-adapter
new ParseServer({
  auth: {
    firebase: "firebase-auth-adapter"
  }
})
```

That's easy enough to implement, what do you think?

**Comment by mockitor** [comment]
i could not find any simple example to implement auth adapter. instead of i've removed firebase-admin-sdk. Firebase Id token is verified against Google Public key using jsonwebtoken jwt library. only jsonwebtoken library is extra dependency now.

**Comment by flovilmart** [comment]
The point was that this auth adapter can and should be distributed as it's own npm package (as we do for the push adapters for example)

**Comment by flovilmart** [comment]
Closing as we're prefer to see that distributed as it's own node module, if you need help on that ping me :)

**Comment by ranhsd** [comment]
Hi @flovilmart- I can create it as external module. Can you please create a repository under parse-server-modules? I will find some time to do it later this week or next week.

@mockitor - I will try to use your code first and will let you know if there are issues.

Thanks.

**Pull request closed without merging**